File: provision/generator/encoding.cpp
#include "assert.h"
#include "encoding.h"
#include <string.h>


void Encoder::init(uint8_t *data, size_t size)
{
    _data = data;
    _size = size;
    reset();
}

void Encoder::reset()
{
    _offset = 0;
}

int Encoder::addUint8(uint8_t in)
{
    ASSERT(_data, return 1, "Invalid pointer");
    ASSERT(_size > 0, return 2, "Invalid size");
    ASSERT((_offset + 2) <= _size, return 3, "Buffer at end");
    _data[_offset++] = Type_UInt8;
    _data[_offset++] = in;
    return 0;
}

int Encoder::getUint8(uint8_t &out)
{
    ASSERT(_data, return 1, "Invalid pointer");
    ASSERT(_size > 0, return 2, "Invalid size");
    ASSERT((_offset + 2) <= _size, return 3, "Buffer at end");
    ASSERT(Type_UInt8 == _data[_offset++], return 4, "Invalid type");

    out = _data[_offset++];
    return 0;
}

int Encoder::addUint16(uint16_t in)
{
    ASSERT(_data, return 1, "Invalid pointer");
    ASSERT(_size > 0, return 2, "Invalid size");
    ASSERT((_offset + 3) <= _size, return 3, "Buffer at end");
    _data[_offset++] = Type_UInt16;
    _data[_offset++] = (in >> 8) & 0xff;
    _data[_offset++] = in & 0xff;
    return 0;
}

int Encoder::getUint16(uint16_t &out)
{
    ASSERT(_data, return 1, "Invalid pointer");
    ASSERT(_size > 0, return 2, "Invalid size");
    ASSERT((_offset + 3) <= _size, return 3, "Buffer at end");
    ASSERT(Type_UInt16 == _data[_offset++], return 4, "Invalid type");
    const uint8_t *p = &_data[_offset];
    out = (p[0] << 8) + p[1];
    _offset += 2;
    return 0;
}

int Encoder::addUint32(uint32_t in)
{
    ASSERT(_data, return 1, "Invalid pointer");
    ASSERT(_size > 0, return 2, "Invalid size");
    ASSERT((_offset + 3) <= _size, return 3, "Buffer at end");
    _data[_offset++] = Type_UInt32;
    _data[_offset++] = (in >> 24) & 0xff;
    _data[_offset++] = (in >> 16) & 0xff;
    _data[_offset++] = (in >> 8) & 0xff;
    _data[_offset++] = in & 0xff;
    return 0;
}

int Encoder::addInt32(int32_t in)
{
    return (in < 0) ? addUint32(0x80000000 | (-in)) : addUint32(in);
}

int Encoder::getUint32(uint32_t &out)
{
    ASSERT(_data, return 1, "Invalid pointer");
    ASSERT(_size > 0, return 2, "Invalid size");
    ASSERT((_offset + 3) <= _size, return 3, "Buffer at end");
    ASSERT(Type_UInt32 == _data[_offset++], return 4, "Invalid type");
    const uint8_t *p = &_data[_offset];
    out = (p[0] << 24) + (p[1] << 16) + (p[2] << 8) + p[3];
    _offset += 4;
    return 0;
}


int Encoder::addArray(const uint8_t *in, size_t size)
{
    ASSERT(_data, return 1, "Invalid pointer");
    ASSERT(_size > 0, return 2, "Invalid size");
    ASSERT((_offset + 3) <= _size, return 3, "Buffer at end");
    _data[_offset++] = Type_Array;
    _data[_offset++] = (size >> 8) & 0xff;
    _data[_offset++] = (size) & 0xff;
    memcpy(&_data[_offset], in, size);
    _offset += size;
    return 0;
}

int Encoder::getArray(uint8_t *out, size_t max_size, size_t &size)
{
    size = 0;
    ASSERT(_data, return 1, "Invalid pointer");
    ASSERT(_size > 0, return 2, "Invalid size");
    ASSERT((_offset + 3) <= _size, return 3, "Buffer at end");
    ASSERT(Type_Array == _data[_offset++], return 4, "Invalid type");

    const uint8_t *p = &_data[_offset];
    size_t sz = (p[0] << 8) + p[1];
    ASSERT((sz <= max_size) && (_offset + 2 + sz <= _size), return 5, "Invalid size");
    memcpy(out, &p[2], sz);
    size = sz;
    _offset += 2 + sz;
    return 0;
}

int Encoder::addString(const char *in)
{
    ASSERT(_data, return 1, "Invalid pointer");
    ASSERT(_size > 0, return 2, "Invalid size");
    size_t size = strlen(in);
    return addArray((uint8_t *)in, size + 1);
}

int Encoder::getString(char *out, size_t max_size, size_t &size)
{
    ASSERT(max_size > 0, return 1, "Invalid size");
    int err = this->getArray((uint8_t *)out, max_size - 1, size);
    ASSERT(!err, return err, "Array encoding error");
    out[size] = 0;
    return 0;
}

File: provision/modules/encoding.py

class Encoder(object):
    UINT8 = 1
    UINT16 = 2
    UINT32 = 3
    ARRAY = 4

    def __init__(self, data = None):
        if data is None:
            self.data = bytearray()
        else:
            self.data = data
        self.offset = 0


    def addType(self, x):
        self.data.extend(int(x).to_bytes(1, 'big'))

    def getType(self):
        t = int.from_bytes(self.data[self.offset:self.offset + 1], byteorder='big')
        self.offset = self.offset + 1
        return t


    def addUint8(self, x):
        self.addType(Encoder.UINT8)
        self.data.extend(int(x).to_bytes(1, 'big'))

    def getUint8(self):
        assert(Encoder.UINT8 == self.getType())
        x = int.from_bytes(self.data[self.offset:self.offset + 1], byteorder='big')
        self.offset = self.offset + 1
        return x


    def addUint16(self, x):
        self.addType(Encoder.UINT16)
        self.data.extend(int(x).to_bytes(2, 'big'))

    def getUint16(self):
        assert(Encoder.UINT16 == self.getType())
        x = int.from_bytes(self.data[self.offset:self.offset + 2], byteorder='big')
        self.offset = self.offset + 2
        return x


    def addUint32(self, x):
        self.addType(Encoder.UINT32)
        self.data.extend(int(x).to_bytes(4, 'big'))

    def getUint32(self):
        assert(Encoder.UINT32 == self.getType())
        x = int.from_bytes(self.data[self.offset:self.offset + 4], byteorder='big')
        self.offset = self.offset + 4
        return x

    def getInt32(self):
        x = self.getUint32()
        if x & 0x80000000:
            return -(0x7fffffff & x);
        else:
            return x

    def addArray(self, x):
        # a = x.encode()
        if x is None:
            x = bytearray()
        self.addType(Encoder.ARRAY)
        self.data.extend(int(len(x)).to_bytes(2, 'big'))
        self.data.extend(x)

    def getArray(self):
        assert(Encoder.ARRAY == self.getType())
        z = int.from_bytes(self.data[self.offset:self.offset + 2], byteorder='big')
        a = self.data[self.offset + 2:self.offset + 2 + z]
        self.offset += (2 + z)
        return bytearray(a)


    def addString(self, x):
        if x is None:
            return self.addArray(x)
        else:
            return self.addArray(x.encode('utf-8'))

    def getString(self):
        a = self.getArray()
        return a.decode('utf-8')

    def serialize(self):
        return self.data

    @staticmethod
    def hex(arr):
        return str(len(arr)) + ':' + ' '.join(format(x, '02x') for x in arr)
